Refactor inplace edit storage

furinvader opened the pull request:
Basically, move the whole save and update functionality to an appropriate service.

furinvader committed:
Move update and save methods to inplaceEditStorage service

furinvader committed:
Reset some states after successful WP update

furinvader committed:
Update resource property after WP update

furinvader committed:
Return promise on error

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://furinvader-features-refactoring-16364-storage.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://furinvader-features-refactoring-16364-storage.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

furinvader committed:
Catch promise rejection of updateWorkPakageForm method

oliverguenther commented:
Some of the failing tests stem from the changed implementation of combined commenting and editing.

Previously, when editing a comment AND a field, hitting save on either of the two fields would save both fields. Did this change in the current specification?

furinvader committed:
Remove ApiHelper dependency

furinvader committed:
Reset errors after rejection

furinvader commented:
The previous behaviour should not have changed.
Something else is probably causing this.

furinvader committed:
Remove submissionPromises reference from WP comment directive

furinvader committed:
Implement add comment method in inplaceEditStorage service

furinvader committed:
Implement multi storage service

For saving multiple inplace forms at once in
a certain order.

furinvader committed:
Show only one validation error

Else, for every existing field, the same error
would be shown.

furinvader committed:
Use storage save method in edit-actions directive

furinvader committed:
Update form on type update

furinvader committed:
Use $scope instead of $rootScope in a controller

Preventing multiple error messages to show
up after validation fails.

furinvader committed:
Reset errors after promises are resolved

oliverguenther commented:
There's a discrepancy in the handling of API errors (e.g., concurrent modification) and field errors (e.g., subject too long).

https://github.com/furinvader/openproject/pull/2 should address this issue.

Currently, you're handling field-specific failures in the catch here: https://github.com/opf/openproject/pull/3753/files#diff-056c28bed7512e7262259a25cf7791f8R182

But the `updateWorkPackageForm` function also handled API errors before, which are now missed as you simply check for field errors.

I've also skipped early evaluation of the promise in other fields when they're not active, otherwise you'll evaluate the same promise for all fields, even when they're not active (you may have a better idea here).

There's still an error in the new work package creation, since it still uses the old promises. Let's talk about this one directly, since I'm not sure If you want to drop the old implementation entirely, or whether the force edit mode will still require some other steps I do not see.

oliverguenther commented on frontend/app/components/inplace-edit/directives/edit-pane/edit-pane.directive.js:
```diff
@@ -88,110 +88,33 @@ function inplaceEditorEditPane(EditableFieldsState, FocusHelper, $timeout) {
-  this.handleFailure = function(e, submit) {
-    afterError();
-    EditableFieldsState.errors = {
-      _common: ApiHelper.getErrorMessages(e)
```
This is the section that handled API errors previously. I've re-introduced it in my PR.

oliverguenther commented on frontend/app/components/inplace-edit/directives/edit-pane/edit-pane.directive.js:
```diff
@@ -251,14 +165,30 @@ function InplaceEditorEditPaneController($scope, $element, $location, $timeout,
+
+  $scope.$on('inplaceEditMultiStorage.save.workPackage', function (event, promise) {
+    promise.then(function() {
+      $location.hash(null);
```
If you're not skipping the promise for non-active fields, you will essentially (try to) focus on all of them. Not sure what will happen in that case. There's also the case of forced edit, where this should not happen.

furinvader replied:
Right, I did not see, that those things would only make sense, if you actually click or somehow make the field active.

furinvader committed:
Evaluate field actions in submit method

furinvader committed:
Reintroduce API error handling

oliverguenther commented on frontend/app/components/inplace-edit/services/inplace-edit-storage.service.js:
```diff
@@ -0,0 +1,103 @@
+        .catch(function (errors) {
+          EditableFieldsState.errors = {
+            _common: ApiHelper.getErrorMessages(errors)
+          };
```
You still must reject the promise here or otherwise, the error handler won't fire. This is causing the `features/work_packages/update.feature` failure on travis.

furinvader committed:
Move notifications to multi storage service

furinvader committed:
Update workPackage.form after form is loaded

furinvader committed:
Show no confirmation message if WP is being created

furinvader committed:
Handle API errors

furinvader committed:
Remove EditableFieldsState.save method

Also, al occurrences of submissionPromises
as they are replaced by the multiStore.stores
attribute.

furinvader committed:
Fix API error rejection

furinvader committed:
Remove unused dependencies

furinvader committed:
Set resource of inplace-edit form if not set before

oliverguenther commented:
I've seen a flickering cuke that seems to be related to notifications. I can't quite rule out that its from this PR, but I can't reproduce it anymore locally. It may as well originate from my notifications PR.
